cmd/kubeadm/app/preflight/checks_linux.go

```
@@ -88,5 +88,13 @@ func addExecChecks(checks []Checker, execer utilsexec.Interface, k8sVersion stri
 
 	// kubelet requires mount to be present in PATH for in-tree volume plugins.
 	checks = append(checks, InPathCheck{executable: "mount", mandatory: true, exec: execer})
+
+	// kubeadm requires cp to be present in PATH for copying etcd directories.
+	checks = append(checks, InPathCheck{executable: "cp", mandatory: true, exec: execer})
+
+	// kubeadm requires the following binaries to be present in PATH for managing init systems.
+	checks = append(checks, InPathCheck{executable: "rc-service", mandatory: true, exec: execer})
+	checks = append(checks, InPathCheck{executable: "rc-update", mandatory: true, exec: execer})
+	checks = append(checks, InPathCheck{executable: "systemctl", mandatory: true, exec: execer})
```

openrc is used instead of systemd on alpine linux. i think we handled that in the initsystem package

---

so this makes the check tricky on unix - error if both init system control binaries are not found.

---

> openrc is used instead of systemd on alpine linux. i think we handled that in the initsystem package.

You are right. Let me remove these checks.

https://github.com/kubernetes/kubernetes/blob/69ab91a5c59617872c9f48737c64409a9dec2957/cmd/kubeadm/app/util/initsystem/initsystem_unix.go#L155

---

And the initsystem has been checked in `ServiceCheck`

---

Should I update the function as below?
```go
diff --git a/cmd/kubeadm/app/util/initsystem/initsystem_unix.go b/cmd/kubeadm/app/util/initsystem/initsystem_unix.go
index 04c21772127..93ba9b9becb 100644
--- a/cmd/kubeadm/app/util/initsystem/initsystem_unix.go
+++ b/cmd/kubeadm/app/util/initsystem/initsystem_unix.go
@@ -160,6 +160,13 @@ func GetInitSystem() (InitSystem, error) {
        }
        _, err = exec.LookPath("openrc")
        if err == nil {
+               binaries := []string{"rc-service", "rc-update"}
+               for _, binary := range binaries {
+                       _, err = exec.LookPath(binary)
+                       if err != nil {
+                               return nil, errors.Wrapf(err, "openrc detected, but missing required binary: %s", binary)
+                       }
+               }
                return &OpenRCInitSystem{}, nil
        }
```

---

or 
```go
diff --git a/cmd/kubeadm/app/preflight/checks_linux.go b/cmd/kubeadm/app/preflight/checks_linux.go
index bc1ce2683d6..bef3e6d4d2f 100644
--- a/cmd/kubeadm/app/preflight/checks_linux.go
+++ b/cmd/kubeadm/app/preflight/checks_linux.go
@@ -25,6 +25,7 @@ import (
        "github.com/pkg/errors"

        utilversion "k8s.io/apimachinery/pkg/util/version"
+       "k8s.io/kubernetes/cmd/kubeadm/app/util/initsystem"
        system "k8s.io/system-validators/validators"
        utilsexec "k8s.io/utils/exec"
 )
@@ -91,5 +92,18 @@ func addExecChecks(checks []Checker, execer utilsexec.Interface, k8sVersion stri

        // kubeadm requires cp to be present in PATH for copying etcd directories.
        checks = append(checks, InPathCheck{executable: "cp", mandatory: true, exec: execer})
+
+       // kubeadm requires the following binaries to be present in PATH for managing init systems.
+       initSystem, err := initsystem.GetInitSystem()
+       if err == nil {
+               switch initSystem.(type) {
+               case *initsystem.SystemdInitSystem:
+                       checks = append(checks, InPathCheck{executable: "systemctl", mandatory: true, exec: execer})
+               case *initsystem.OpenRCInitSystem:
+                       checks = append(checks, InPathCheck{executable: "rc-service", mandatory: true,
 exec: execer})
+                       checks = append(checks, InPathCheck{executable: "rc-update", mandatory: true,
exec: execer})
+               }
+       }
+
        return checks
 }
```

---

GetInitSystem uses the lookpath approach, so i think it make sense to have all the lookpath checks there.
i.e. first option from your comments above.

we already have `func (sc ServiceCheck) Check() (warnings, errorList []error) {`, which calls GetInitSystem so it would fail early on missing binaries for openrc. it already fails for systemctl.

---

updated.